### PR TITLE
Account Version for both account import and export

### PR DESCRIPTION
**Pull request by EvanJRichard**
## Summary
`AccountImport`s now have a `version` field. If a user imports interactively, or if the provided data otherwise has no version, the version is 1. 

This builds off of and includes #3222

## Testing Plan
```
➜  ironfish-cli git:(evan/importaccountversion) ✗ yarn start wallet:export importtest
yarn run v1.22.19
$ yarn build && yarn start:js wallet:export importtest
$ tsc -b
$ cross-env OCLIF_TS_NODE=0 IRONFISH_DEBUG=1 node --expose-gc --inspect=:0 --inspect-publish-uid=http --enable-source-maps bin/run wallet:export importtest
{
   "id": "165a8021-e719-4857-826b-7e4455518803",
   "name": "importtest",
   "spendingKey": "7bb41f83fd156535f9adcf1d88d9322a1c4d6baf6cc03bbcab962f901e316d62",
   "incomingViewKey": "090a2a6fa0e4f67923e1802aaa59dbe30229916307d81c5cd9c70f7853e65c06",
   "outgoingViewKey": "02dea4335761ed93e88c69251c323b3f1e22632efadfc1ccf852bc6a0d6d3bb5",
   "publicAddress": "e0555fa13f968e598ae9afd05d41cf17ae91c91423f3f07091871ed1a08d7ca6",
   "version": 1
}
✨  Done in 2.23s.
➜  ironfish-cli git:(evan/importaccountversion) yarn start wallet:export importtest testfile
yarn run v1.22.19
$ yarn build && yarn start:js wallet:export importtest testfile
$ tsc -b
$ cross-env OCLIF_TS_NODE=0 IRONFISH_DEBUG=1 node --expose-gc --inspect=:0 --inspect-publish-uid=http --enable-source-maps bin/run wallet:export importtest testfile
✨  Done in 1.16s.
➜  ironfish-cli git:(evan/importaccountversion) ✗ yarn start wallet:import testfile
yarn run v1.22.19
$ yarn build && yarn start:js wallet:import testfile
$ tsc -b
$ cross-env OCLIF_TS_NODE=0 IRONFISH_DEBUG=1 node --expose-gc --inspect=:0 --inspect-publish-uid=http --enable-source-maps bin/run wallet:import testfile
Account already exists with the name importtest
✨  Done in 1.17s.
```

Migration upgrade
```
Applying 1 migrations:
  021-add-version-to-accounts

Running 021-add-version-to-accounts...
Gathering accounts for migration

  Migrating account importtest
  Completed migration for account importtest

  Migrating account evanironfish
  Completed migration for account evanironfish
Successfully applied 1 migrations
WebSocket server started at :::9033
Connected to the Iron Fish network
```

## Breaking Change

Is this a breaking change? If yes, add notes below on why this is breaking and
what additional work is required, if any.

```
[ ] Yes
[X] No
```

### CODE REVIEW

File: ironfish/src/wallet/walletdb/accountValue.ts [hunk 1]
```diff
@@ -25,6 +26,7 @@ export class AccountValueEncoding implements IDatabaseEncoding<AccountValue> {
     bw.writeBytes(Buffer.from(value.incomingViewKey, 'hex'))
     bw.writeBytes(Buffer.from(value.outgoingViewKey, 'hex'))
     bw.writeBytes(Buffer.from(value.publicAddress, 'hex'))
+    bw.writeU8(value.version)
```
Comment (jowparks) [inline ironfish/src/wallet/walletdb/accountValue.ts:29]: 256 versions seems safe in terms of upper limit, but I think 16bit is even safer

Comment (NullSoldier) [reply ironfish/src/wallet/walletdb/accountValue.ts:29]: Don't you need version to be 1st to descriminate deserialization on it?

File: ironfish/src/wallet/walletdb/accountValue.ts [hunk 2]
```diff
@@ -37,6 +39,10 @@ export class AccountValueEncoding implements IDatabaseEncoding<AccountValue> {
     const incomingViewKey = reader.readBytes(KEY_LENGTH).toString('hex')
     const outgoingViewKey = reader.readBytes(KEY_LENGTH).toString('hex')
     const publicAddress = reader.readBytes(PUBLIC_ADDRESS_LENGTH).toString('hex')
+    let version = 1
+    if (reader.left() !== 0) {
+      version = reader.readU8()
+    }
```
[the remark this answers is not in the document]
Comment (NullSoldier) [reply ironfish/src/wallet/walletdb/accountValue.ts:45]: I can show you how to write a quick one, but migration 17 is pretty easy to copy to do what you want

[the remark this answers is not in the document]
Comment (EvanJRichard) [reply ironfish/src/wallet/walletdb/accountValue.ts:45]: @NullSoldier I'd certainly rather a migration than this, can you point me to migration 17 please?

File: ironfish/src/wallet/walletdb/accountValue.ts (hunk 1 above)
Comment (EvanJRichard) [reply ironfish/src/wallet/walletdb/accountValue.ts:29]: @jowparks sounds good!
@NullSoldier version last is doable but insanely janky compared to version first, combined with your "let's just do a migration" comment I'll happily move it to the front

File: ironfish/src/wallet/walletdb/accountValue.ts [hunk 3]
```diff
@@ -56,6 +63,7 @@ export class AccountValueEncoding implements IDatabaseEncoding<AccountValue> {
     size += KEY_LENGTH
     size += KEY_LENGTH
     size += PUBLIC_ADDRESS_LENGTH
+    size += 1 // version
```
Comment (EvanJRichard) [inline ironfish/src/wallet/walletdb/accountValue.ts]: @ anyone : where would be a good place to define `VERSION_LENGTH`? Just at the top of `accountValue.ts` here?

File: ironfish/src/wallet/walletdb/accountValue.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (EvanJRichard) [reply ironfish/src/wallet/walletdb/accountValue.ts:45]: found it, noting it for myself #3257

File: ironfish/src/wallet/walletdb/accountValue.ts (hunk 3 above)
Comment (EvanJRichard) [reply ironfish/src/wallet/walletdb/accountValue.ts]: doing that

File: ironfish/src/wallet/walletdb/accountValue.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (EvanJRichard) [reply ironfish/src/wallet/walletdb/accountValue.ts]: Decided after our engineering meeting on Friday that it's worth it to keep the version in the account struct. So, spent some time today trying to learn the migration stuff - worried that I'm not understanding it properly, having a little trouble, so pushed my WIP and going to circle back to it tomorrow.

File: ironfish/src/migrations/data/021-add-version-to-accounts.ts [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+        name: account.name,
+        spendingKey: account.spendingKey,
+        version: account.version,
+      }
+      await node.wallet.importAccount(importRequest as AccountImport)
+    }
+  }
+
+  async backward(node: IronfishNode, db: IDatabase): Promise<void> {
+    // TODO something that undoes the migration? Some migrations don't implement this though?
```
Comment (jowparks) [inline ironfish/src/migrations/data/021-add-version-to-accounts.ts:56]: Don't need to worry about this.

File: ironfish/src/migrations/data/021-add-version-to-accounts.ts [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+  async forward(
+    node: IronfishNode,
+    _db: IDatabase,
+    _tx: IDatabaseTransaction | undefined,
+    logger: Logger,
+  ): Promise<void> {
+    const accounts = []
+    // TODO: does this serialize/deserialize correctly? it assumes Version already exists when it in fact does not.
+    logger.debug(`Loading accounts from wallet db...`)
+    for await (const accountValue of node.wallet.walletDb.loadAccounts()) {
```
Comment (NullSoldier) [inline ironfish/src/migrations/data/021-add-version-to-accounts.ts:31]: you cannot do this because the walletDB has the old spec

**Comment by EvanJRichard** [comment]
Migration written and successfully ran locally. Would appreciate extra scrutiny on all walletdb-touching aspects, touching the db makes me nervous.

File: ironfish/src/wallet/walletdb/walletdb.ts [hunk 6]
```diff
@@ -143,7 +143,7 @@ export class WalletDB {
     })
 
     this.accounts = this.db.addStore({
```
Comment (jowparks) [inline ironfish/src/wallet/walletdb/walletdb.ts:145]: nice

File: ironfish/src/rpc/routes/wallet/importAccount.ts [hunk 7]
```diff
@@ -22,6 +22,7 @@ export const ImportAccountRequestSchema: yup.ObjectSchema<ImportAccountRequest>
       .object({
         name: yup.string().defined(),
         spendingKey: yup.string().defined(),
+        version: yup.number().defined(),
```
Comment (jowparks) [inline ironfish/src/rpc/routes/wallet/importAccount.ts:25]: The version is possibly undefined from old account exports, right?

File: ironfish-cli/src/commands/wallet/import.ts [hunk 8]
```diff
@@ -110,7 +110,7 @@ export class ImportCommand extends IronfishCommand {
       const name = await CliUx.ux.prompt('Enter a new account name', {
         required: true,
       })
-      return { name, spendingKey }
+      return { name, spendingKey, version: 1 }
```
Comment (jowparks) [inline ironfish-cli/src/commands/wallet/import.ts:113]: Can we create a constant somewhere we can bump when version changes occur and reference it here?

**Review by jowparks — APPROVED** [review]
LGTM, some small comments. Once you address them, be sure to retest and paste output in PR comment. Also can you neuter the shell output to be just the relevant lines? It is tough to read without coloring what is actually relevant.

File: ironfish/src/rpc/routes/wallet/importAccount.ts (hunk 7 above)
Comment (EvanJRichard) [reply ironfish/src/rpc/routes/wallet/importAccount.ts:25]: Right, so we add it in the CLI if none is defined.

Comment (jowparks) [reply ironfish/src/rpc/routes/wallet/importAccount.ts:25]: Ahh cool sounds good

**Comment by EvanJRichard** [comment]
Another successful migration: 
```
Applying 1 migrations:
  021-add-version-to-accounts

Running 021-add-version-to-accounts...
Gathering accounts for migration

  Migrating account importtest
  Completed migration for account importtest

  Migrating account evanironfish
  Completed migration for account evanironfish
```

**Comment by EvanJRichard** [comment]
Another successful export: 
`yarn start wallet:export importtest --json`
yields
```
{
    "version": 1,
    "name": "importtest",
    "spendingKey": "7bb41f83fd156535f9adcf1d88d9322a1c4d6baf6cc03bbcab962f901e316d62",
    "incomingViewKey": "090a2a6fa0e4f67923e1802aaa59dbe30229916307d81c5cd9c70f7853e65c06",
    "outgoingViewKey": "02dea4335761ed93e88c69251c323b3f1e22632efadfc1ccf852bc6a0d6d3bb5",
    "publicAddress": "e0555fa13f968e598ae9afd05d41cf17ae91c91423f3f07091871ed1a08d7ca6"
}
```

**Pull request merged**